airbyte-cdk/python/airbyte_cdk/utils/print_buffer.py

```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+import sys
+import time
+from io import StringIO
+from threading import RLock
+from types import TracebackType
+from typing import Optional
+
+
+class PrintBuffer:
+    def __init__(self, flush_interval: float = 0.1):
+        self.buffer = StringIO()
+        self.flush_interval = flush_interval
+        self.last_flush_time = time.monotonic()
+        self.lock = RLock()
+
+    def write(self, message: str) -> None:
+        with self.lock:
+            self.buffer.write(message)
+            current_time = time.monotonic()
+            if (current_time - self.last_flush_time) >= self.flush_interval:
```

Comment:
Am I correct to assume that it flushes only on write OR manually, but not automatically on timer?

Does this allow us to footgun ourselves if we write once and forget about the buffer without a manual flush? Hypothetically no, because we have `__exit__`?

Reply:
> Am I correct to assume that it flushes only on write OR manually, but not automatically on timer?

Yep, since all things are done in 1 thread, it flushes only on write method. Automatically on timer is described here: https://github.com/airbytehq/airbyte-internal-issues/issues/8860#issuecomment-2255489148, but python is not very good at this.

> Does this allow us to footgun ourselves if we write once and forget about the buffer without a manual flush?

No, we'll always flush on context __exit__.